crates/wit-component/src/linking.rs

```diff
@@ -0,0 +1,1383 @@
+        &crate::base_producers().raw_custom_section(),
+    ));
+
+    let module = module.finish();
+    wasmparser::validate(&module)?;
+
+    Ok(module)
+}
+
+/// Find the library which exports the specified function or global address.
+fn find_offset_exporter<'a>(
+    name: &str,
+    exporters: &HashMap<&ExportKey, (&'a str, &'a Export<'a>)>,
+) -> Result<(&'a str, &'a Export<'a>)> {
+    let export = ExportKey {
+        name,
+        ty: Type::Global(GlobalType {
+            ty: ValueType::I32,
+            mutable: false,
+        }),
+    };
+
+    exporters
+        .get(&export)
+        .copied()
+        .ok_or_else(|| anyhow!("unable to find {export:?} in any library"))
+}
+
+/// Find the library which exports the specified function.
+fn find_function_exporter<'a>(
+    name: &str,
+    ty: &FunctionType,
+    exporters: &HashMap<&ExportKey, (&'a str, &'a Export<'a>)>,
+) -> Result<(&'a str, &'a Export<'a>)> {
+    let export = ExportKey {
+        name,
+        ty: Type::Function(ty.clone()),
+    };
+
+    exporters
+        .get(&export)
+        .copied()
+        .ok_or_else(|| anyhow!("unable to find {export:?} in any library"))
+}
+
+/// Analyze the specified library metadata, producing a symbol-to-library-name map of exports.
+fn resolve_exporters<'a>(
+    metadata: &'a [Metadata<'a>],
+) -> Result<HashMap<&'a ExportKey<'a>, Vec<(&'a str, &'a Export<'a>)>>> {
+    let mut exporters = HashMap::<_, Vec<_>>::new();
+    for metadata in metadata {
+        for export in &metadata.exports {
+            exporters
+                .entry(&export.key)
+                .or_default()
+                .push((metadata.name, export));
+        }
+    }
+    Ok(exporters)
+}
+
+/// Match up all imported symbols to their corresponding exports, reporting any missing or duplicate symbols.
+fn resolve_symbols<'a>(
+    metadata: &'a [Metadata<'a>],
+    exporters: &'a HashMap<&'a ExportKey<'a>, Vec<(&'a str, &'a Export<'a>)>>,
+) -> (
+    HashMap<&'a ExportKey<'a>, (&'a str, &'a Export<'a>)>,
+    Vec<(&'a str, Export<'a>)>,
+    Vec<(&'a str, &'a ExportKey<'a>, &'a [(&'a str, &'a Export<'a>)])>,
+) {
+    // TODO: consider weak symbols when checking for duplicates
+
+    let function_exporters = exporters
+        .iter()
+        .filter_map(|(export, exporters)| {
+            if let Type::Function(_) = &export.ty {
+                Some((export.name, (export, exporters)))
+            } else {
+                None
+            }
+        })
+        .collect::<HashMap<_, _>>();
+
+    let mut resolved = HashMap::new();
+    let mut missing = Vec::new();
+    let mut duplicates = Vec::new();
+
+    let mut triage = |metadata: &'a Metadata, export: Export<'a>| {
+        if let Some((key, value)) = exporters.get_key_value(&export.key) {
+            match value.as_slice() {
+                [] => unreachable!(),
+                [exporter] => {
+                    resolved.insert(*key, *exporter);
```

Comment:
Could this assert that the insertion didn't replace anything?

Reply:
In this case, we actually expect it may replace something since multiple libraries may import the same symbol.  I've added a comment to note that.  I've also added assertions to a few places where we don't expect replacement.